### ISSUE TITLE
Iteration 3: Multiple applications and headless nodes with dynamic network membership

### ISSUE THREAD

**Issue by gnulib** [trust-net/dag-lib-go#26]
This is the third iteration for MVP defined in #21, as following:
### Objective
_As a GO application developer, I can build and deploy multiple native GO applications using DLT stack library, so that each application node can join the network dynamically and sync with its own shard's transactions across the network even if application nodes are not directly connected as peers._

### Assumptions
* ~all nodes are initialized and connected to the network before first transaction is submitted (node's do not join an existing network with previous history)~  
nodes can join the network at any time after transaction processing has started
* ~nodes stay online and connected to the network for duration/life of the experiment~   
nodes can shutdown/leave the network, and join back dynamically
* ~nodes do not need to persist and restore transaction history or consensus state~   
node sync will be "full resync", as long as there is some application node with transactions history the network sync will work for that application's shard
* all nodes are completely reliable (no fault tolerance required) and honest (no competing transactions are submitted)
* different nodes can run different applications, and can even run without application (headless node)

### Acceptance Criteria
* p2p: discovery and connection, basic handshake for protocol version compatibility
* endorsement: pass through, no anchoring and submission, ~no peer sync~ peer sync at sharding layer (~all nodes online at genesis~ nodes can join/leave network dynamically)
* shard: app transaction filtering based on shard membership, supports headless node
* privacy: pass through (no payload encryption/decryption/exchange)
* ~consensus: basic handshake for app version compatibility (should this actually be in shard layer?), no peer sync (all nodes online at genesis), no conflict/concurrency resolution, no persistence~   
consensus: full node sync upon startup/joining a network, no conflict/concurrency resolution, no persistence
* demo app: a simple network counter app that can register specific shard with DLT stack and send increment and decrement counter op transactions once all nodes are setup and connected (i.e. all nodes online at genesis). All application nodes within same shard should be consistent in counter values.

**Comment by gnulib** [comment]
## What does consensus mean?
There are two levels of “consensus”, the headless node level consensus and the application shard level consensus. They both have specific purpose:
* node level consensus deals with endorsement, anchoring etc. — basically its the meta-data consensus about transaction history for submitters and application shards. This consensus does not require knowledge about a transaction’s specific operations/logic, it just cares about meta data (who/how/where/when) for each transaction
* application level consensus deals with actual world state view from application shard’s point of view. Hence, this consensus requires knowledge about each transaction’s specific operation/logic. This can be accomplished by providing a “World State” API to application layer, and let the application process transaction and update the world state for that application’s shard.

**Comment by gnulib** [comment]
## How will application sync work?
Application nodes may not directly be connected with other application nodes from same shard. In that case, how will an application node get synced up with the shard?
* one option is to have the sharding layer maintain full transaction history of all application shards in the network. It may not process them all, but will maintain full transaction DAGs as it witnesses on the network. Then, when application node connects, it will sync at the sharding layer, and process the registered application’s transactions during sync, to build/update the world state for its shard
* other option is to “probe" and “discover" another application node on the network, and initiate sync with that node directly. Question then is, should this probe and discovery result in a new direct p2p layer connection with the application peer on network? Its certainly possible to establish connection if nodeID is known (e.g. as a response to probe message), however, not all nodes may be able to accept incoming connection request (e.g. NAT). Hence, this approach has sever limitations, and cannot be relied upon to work all the time!!!
* A third option is to “probe” and “flood” a sync “broadcast”, i.e., let an application node joining the network send out a sync initiation request on the network that gets flooded to all nodes, and let the application nodes respond back with sync responses that get flooded to the network. This method seems not scalable at world scale, especially when we envision number of application nodes << network size, which means most of the sync traffic will be noise and useless.

It seems, the first option is the most viable option. We can build a sync protocol for this approach as following:
* let an application node joining the network initiate a shard level sync with its connected peers by exchanging their “best” block for the shard
* the peer with “winning” block will send all the “previous” blocks to the other node
* the other node will process received missing blocks and forward to other connected nodes
* note that sync is only initiated by the application node, in other words, shard will only initiate a sync when there is a registered application, and will not initiate a sync for any other application shards
* also, a corollary of above is that if a shard receives a block that it knows nothing about, it will ask the peer for all preceding blocks for that application shard

So, in other words, application level sync will be initiated under two scenarios:
* Application driven: this will be a sync initiated by sharding layer on behalf of it's registered application, and will sync with all connected peers using application provided best block
* Network driven: this will be a sync initiated by sharding layer when it receives a block for any shard (regardless of its own registered application), and when it does not know the parent of that received block. This sync will be specific with the peer node that forwarded the application block.

**Comment by gnulib** [comment]
## Sync of the Shards
We'll need to define new messages for conducting sync of specific shard between connected peers.

**Comment by gnulib** [comment]
## Is it really this complex?
This item has not had any actual development for about a month now. Seems we are over-designing here. Lets consider the facts here:
* issue requirement is for application nodes to be able to sync when they join network
* regardless of whatever consensus and sync mechanism is used, the endorsement layer will keep "full network history" of all transactions for all submitters
* unless we can optimize and eliminate "full network history" at the endorsement layer, till then it makes sense to just do the replay of application specific transactions from endorsement layer when application layer registers

So, it seems, keeping a simple "full network history" at endorsement layer, and replay of application specific transactions for registered shard is the necessary and sufficient scope for this work item.

**Comment by gnulib** [comment]
## Application replay
Endorsement layer will persist ALL transactions in the network. Once an application registers for a specific shard, then endorsement layer will replay transactions for that application shard from its persistent repository of all transactions.

**Comment by gnulib** [comment]
## Network Sync
Now that transaction replay is in place, an app can register anytime (or leave a shard and join a different shard dynamically) and it will be synced up with rest of the apps in that shard. However, once the node shuts down, it will get out of sync with network due to which replay will not have latest transactions.

Hence, for a node to be able to shutdown/restart we need the network sync between nodes working next.
